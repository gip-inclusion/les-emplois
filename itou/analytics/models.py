import uuid

from django.db import models
from django.utils import timezone

from itou.common_apps.address.departments import DEPARTMENTS
from itou.users.enums import UserKind


class DatumCode(models.TextChoices):
    # Employee record - Base
    EMPLOYEE_RECORD_COUNT = "ER-001", "FS totales"
    EMPLOYEE_RECORD_DELETED = "ER-002", "FS (probablement) supprimées"
    # Employee record - Lifecycle
    EMPLOYEE_RECORD_PROCESSED_AT_FIRST_EXCHANGE = "ER-101", "FS intégrées (0000) au premier retour"
    EMPLOYEE_RECORD_WITH_ERROR_AT_FIRST_EXCHANGE = "ER-102", "FS avec une erreur au premier retour"
    EMPLOYEE_RECORD_WITH_ERROR_3436_AT_FIRST_EXCHANGE = "ER-102-3436", "FS avec une erreur 3436 au premier retour"
    EMPLOYEE_RECORD_WITH_AT_LEAST_ONE_ERROR = "ER-103", "FS ayant eu au moins un retour en erreur"
    # Approval - Base
    APPROVAL_COUNT = "AP-001", "PASS IAE total"
    APPROVAL_CANCELLED = "AP-002", "PASS IAE annulés"
    # Approval - PE notification cycle
    APPROVAL_PE_NOTIFY_SUCCESS = "AP-101", "PASS IAE synchronisés avec succès avec pole emploi"
    APPROVAL_PE_NOTIFY_PENDING = "AP-102", "PASS IAE en attente de synchronisation avec pole emploi"
    APPROVAL_PE_NOTIFY_ERROR = "AP-103", "PASS IAE en erreur de synchronisation avec pole emploi"
    APPROVAL_PE_NOTIFY_READY = "AP-104", "PASS IAE prêts à être synchronisés avec pole emploi"
    # Users
    USER_COUNT = "US-001", "Nombre d'utilisateurs"
    USER_JOB_SEEKER_COUNT = "US-011", "Nombre de demandeurs d'emploi"
    USER_PRESCRIBER_COUNT = "US-012", "Nombre de prescripteurs"
    USER_EMPLOYER_COUNT = "US-013", "Nombre d'employeurs"
    USER_LABOR_INSPECTOR_COUNT = "US-014", "Nombre d'inspecteurs du travail"
    USER_ITOU_STAFF_COUNT = "US-015", "Nombre d'administrateurs"
    # API usage
    API_TOTAL_CALLS = "API-001", "API : total d'appels reçus"
    API_TOTAL_UV = "API-002", "API : total de visiteurs uniques"
    API_TOTAL_CALLS_CANDIDATS = "API-003", "API candidats : total d'appels reçus"
    API_TOTAL_UV_CANDIDATS = "API-004", "API candidats : total de visiteurs uniques"
    API_TOTAL_CALLS_GEIQ = "API-005", "API GEIQ : total d'appels reçus"
    API_TOTAL_UV_GEIQ = "API-006", "API GEIQ : total de visiteurs uniques"
    API_TOTAL_CALLS_ER = "API-007", "API FS : total d'appels reçus"
    API_TOTAL_UV_ER = "API-008", "API FS : total de visiteurs uniques"
    API_TOTAL_CALLS_MARCHE = "API-009", "API Le marché : total d'appels reçus"
    API_TOTAL_CALLS_SIAES = "API-010", "API siaes : total d'appels reçus"
    API_TOTAL_UV_SIAES = "API-011", "API siaes : total de visiteurs uniques (par adresse IP)"
    API_TOTAL_CALLS_STRUCTURES = "API-012", "API structures : total d'appels reçus"


class Datum(models.Model):
    """Store an aggregated `value` of the `code` data point for the specified `bucket`."""

    id = models.UUIDField(primary_key=True, default=uuid.uuid4)

    code = models.TextField(choices=DatumCode.choices)
    bucket = models.TextField()
    value = models.IntegerField()  # Integer offers the best balance between range, storage size, and performance

    measured_at = models.DateTimeField(default=timezone.now)  # Not using auto_now_add=True to allow overrides

    class Meta:
        verbose_name_plural = "data"
        unique_together = ["code", "bucket"]
        indexes = [models.Index(fields=["measured_at", "code"])]


class StatsDashboardVisit(models.Model):
    dashboard_id = models.IntegerField(verbose_name="ID tableau de bord Metabase")
    dashboard_name = models.TextField(verbose_name="nom de la vue du tableau de bord")
    department = models.CharField(verbose_name="département", choices=DEPARTMENTS.items(), max_length=3, null=True)
    region = models.TextField(verbose_name="région", null=True)
    current_company_id = models.IntegerField(verbose_name="ID entreprise courante", null=True)
    current_prescriber_organization_id = models.IntegerField(
        verbose_name="ID organisation prescriptrice courante", null=True
    )
    current_institution_id = models.IntegerField(verbose_name="ID institution courante", null=True)
    user_kind = models.TextField(verbose_name="type d'utilisateur", choices=UserKind.choices)
    user_id = models.IntegerField(verbose_name="ID utilisateur")

    measured_at = models.DateTimeField(default=timezone.now)  # Not using auto_now_add=True to allow overrides

    class Meta:
        verbose_name_plural = "visite de tableau de bord"
