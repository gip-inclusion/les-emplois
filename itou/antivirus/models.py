from django.db import models
from django.utils.text import capfirst

from itou.files.models import File


class Scan(models.Model):
    file = models.OneToOneField(File, on_delete=models.CASCADE)
    clamav_signature = models.TextField()
    clamav_completed_at = models.DateTimeField(null=True, verbose_name="analyse ClamAV le")
    clamav_infected = models.BooleanField(null=True, verbose_name="fichier infect√© selon ClamAV")
    comment = models.TextField(blank=True, verbose_name="commentaire")

    class Meta:
        verbose_name = "analyse antivirus"
        verbose_name_plural = "analyses antivirus"

    def __str__(self):
        text = f"{capfirst(self._meta.verbose_name)} {self.file_id}"
        if self.clamav_infected:
            text = f"[VIRUS] {text}"
        return text
