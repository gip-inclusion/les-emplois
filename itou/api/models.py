import secrets

from django.db import models
from django.utils import timezone

from itou.companies.models import Company


def _generate_key():
    return secrets.token_urlsafe()


class CompanyToken(models.Model):
    key = models.CharField(default=_generate_key, unique=True, editable=False)
    label = models.CharField(verbose_name="m√©mo permettant d'identifier l'usage du jeton", max_length=60, unique=True)
    created_at = models.DateTimeField(default=timezone.now)
    companies = models.ManyToManyField(Company, related_name="api_tokens")

    class Meta:
        verbose_name = "jeton d'API entreprise"
        verbose_name_plural = "jetons d'API entreprise"
