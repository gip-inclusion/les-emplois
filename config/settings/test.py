import os


ITOU_ENVIRONMENT = "DEV"
os.environ["ITOU_ENVIRONMENT"] = ITOU_ENVIRONMENT

from .base import *  # noqa: E402,F403


SECRET_KEY = "foobar"
ALLOWED_HOSTS = []

# We *want* to do the same `collectstatic` on the CI than on PROD to catch errors early,
# but we don't want to do it when running the test suite locally for performance reasons.
if not os.getenv("CI", False):
    # `ManifestStaticFilesStorage` (used in base settings) requires `collectstatic` to be run.
    STORAGES["staticfiles"]["BACKEND"] = "django.contrib.staticfiles.storage.StaticFilesStorage"  # noqa: F405

ASP_ITOU_PREFIX = "XXXXX"  # same as in our fixtures
ITOU_PROTOCOL = "http"
ITOU_FQDN = "localhost:8000"

DATABASES["default"]["HOST"] = os.getenv("PGHOST", "127.0.0.1")  # noqa: F405
DATABASES["default"]["PORT"] = os.getenv("PGPORT", "5432")  # noqa: F405
DATABASES["default"]["NAME"] = os.getenv("PGDATABASE", "itou")  # noqa: F405
DATABASES["default"]["USER"] = os.getenv("PGUSER", "postgres")  # noqa: F405
DATABASES["default"]["PASSWORD"] = os.getenv("PGPASSWORD", "password")  # noqa: F405

MAILJET_API_KEY_PRINCIPAL = "API_MAILJET_KEY_PRINCIPAL"
MAILJET_SECRET_KEY_PRINCIPAL = "API_MAILJET_SECRET_PRINCIPAL"

AWS_S3_ENDPOINT_URL = f"http://{os.getenv('CELLAR_ADDON_HOST', 'localhost:9000')}/"
AWS_S3_ACCESS_KEY_ID = "minioadmin"
AWS_S3_SECRET_ACCESS_KEY = "minioadmin"
AWS_STORAGE_BUCKET_NAME = "tests"
