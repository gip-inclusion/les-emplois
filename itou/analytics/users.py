from django.db.models import Count

from itou.users import enums as users_enums, models as users_models

from . import models


def collect_analytics_data(before):
    counts = {
        info["kind"]: info["count"]
        for info in users_models.User.objects.values("kind").annotate(count=Count("pk")).values("kind", "count")
    }
    return {
        models.DatumCode.USER_COUNT: sum(counts.values()),
        models.DatumCode.USER_JOB_SEEKER_COUNT: counts.get(users_enums.KIND_JOB_SEEKER, 0),
        models.DatumCode.USER_PRESCRIBER_COUNT: counts.get(users_enums.KIND_PRESCRIBER, 0),
        models.DatumCode.USER_EMPLOYER_COUNT: counts.get(users_enums.KIND_EMPLOYER, 0),
        models.DatumCode.USER_LABOR_INSPECTOR_COUNT: counts.get(users_enums.KIND_LABOR_INSPECTOR, 0),
        models.DatumCode.USER_ITOU_STAFF_COUNT: counts.get(users_enums.KIND_ITOU_STAFF, 0),
    }
