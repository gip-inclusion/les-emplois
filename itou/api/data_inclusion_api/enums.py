import enum


class StructureTypeStr(str, enum.Enum):
    ORGA = "orga"
    SIAE = "siae"
