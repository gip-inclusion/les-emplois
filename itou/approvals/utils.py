from itou.companies.enums import SIAE_WITH_CONVENTION_KINDS


def get_user_last_accepted_siae_job_application(user):
    if not user.is_job_seeker:
        return None

    # Some candidates may not have accepted job applications
    # Assuming it's the case can lead to issues downstream
    return (
        user.job_applications.accepted()
        .filter(to_company__kind__in=SIAE_WITH_CONVENTION_KINDS)
        .with_accepted_at()
        .order_by("-accepted_at", "-hiring_start_at")
        .first()
    )


def last_hire_was_made_by_siae(user, siae):
    if not user.is_job_seeker:
        return False
    last_accepted_job_application = get_user_last_accepted_siae_job_application(user)
    return last_accepted_job_application and last_accepted_job_application.to_company_id == siae.pk
