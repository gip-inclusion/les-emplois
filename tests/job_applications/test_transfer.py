import pytest
from django.core import mail
from django.core.exceptions import ValidationError
from django.utils import timezone

from itou.eligibility.models import EligibilityDiagnosis
from itou.job_applications.models import JobApplicationWorkflow
from itou.users.enums import UserKind
from tests.companies.factories import CompanyFactory, CompanyWith2MembershipsFactory
from tests.eligibility.factories import EligibilityDiagnosisFactory, EligibilityDiagnosisMadeBySiaeFactory
from tests.job_applications.factories import (
    JobApplicationFactory,
    JobApplicationSentByCompanyFactory,
    JobApplicationSentByPrescriberFactory,
)
from tests.users.factories import JobSeekerFactory
from tests.utils.test import TestCase


class JobApplicationTransferModelTest(TestCase):
    def test_is_in_transferable_state(self):
        # If job application is in ACCEPTED state
        # it can't be transfered
        evil_states = [JobApplicationWorkflow.STATE_ACCEPTED]
        good_states = [
            JobApplicationWorkflow.STATE_NEW,
            JobApplicationWorkflow.STATE_PROCESSING,
            JobApplicationWorkflow.STATE_POSTPONED,
            JobApplicationWorkflow.STATE_REFUSED,
            JobApplicationWorkflow.STATE_CANCELLED,
            JobApplicationWorkflow.STATE_OBSOLETE,
        ]

        for evil_state in evil_states:
            with self.subTest(evil_state):
                job_application = JobApplicationFactory(state=evil_state)
                assert not job_application.is_in_transferable_state

        for good_state in good_states:
            with self.subTest(good_state):
                job_application = JobApplicationFactory(state=good_state)
                assert job_application.is_in_transferable_state

    def test_can_be_transferred(self):
        # Only users in both origin and target SIAE
        # can transfer a job_application
        # (provided it is in correct state)
        origin_company = CompanyFactory(with_membership=True)
        target_company = CompanyFactory(with_membership=True)

        origin_user = origin_company.members.first()
        target_user = target_company.members.first()
        lambda_user = JobSeekerFactory()
        target_company.members.add(origin_user)

        job_application = JobApplicationFactory(to_company=origin_company, state=JobApplicationWorkflow.STATE_ACCEPTED)

        assert origin_user.kind == UserKind.EMPLOYER
        assert target_user.kind == UserKind.EMPLOYER
        assert not job_application.can_be_transferred(target_user, job_application.to_company)
        assert not job_application.can_be_transferred(lambda_user, target_company)
        assert not job_application.can_be_transferred(target_user, target_company)
        assert not job_application.can_be_transferred(origin_user, target_company)

        job_application.state = JobApplicationWorkflow.STATE_PROCESSING

        assert job_application.can_be_transferred(origin_user, target_company)

    def test_transfer_to(self):
        # If all conditions are valid, a user can transfer job applications between SIAE they are member of,
        # provided job application is in an acceptable state.
        # After transfer:
        # - job application is not linked to origin SIAE anymore (only to target SIAE)
        # - eligibility diagnosis is deleted if not created by an authorized prescriber
        origin_company = CompanyFactory(with_membership=True)
        target_company = CompanyFactory(with_membership=True)

        origin_user = origin_company.members.first()
        target_user = target_company.members.first()
        lambda_user = JobSeekerFactory()
        target_company.members.add(origin_user)

        job_application = JobApplicationFactory(
            to_company=origin_company,
            sent_by_authorized_prescriber_organisation=True,
            state=JobApplicationWorkflow.STATE_ACCEPTED,
        )

        # Conditions hould be covered by previous test, but does not hurt (and tests raise)
        with pytest.raises(ValidationError):
            job_application.transfer_to(lambda_user, target_company)
        with pytest.raises(ValidationError):
            job_application.transfer_to(origin_user, origin_company)
        with pytest.raises(ValidationError):
            job_application.transfer_to(target_user, target_company)
        with pytest.raises(ValidationError):
            job_application.transfer_to(origin_user, target_company)

        job_application.state = JobApplicationWorkflow.STATE_PROCESSING
        job_application.transfer_to(origin_user, target_company)
        job_application.refresh_from_db()

        # "Normal" transfer
        assert job_application.to_company == target_company
        assert job_application.state == JobApplicationWorkflow.STATE_NEW
        assert job_application.eligibility_diagnosis is not None

        # Eligibilty diagnosis not sent by authorized prescriber must be deleted
        job_application = JobApplicationSentByCompanyFactory(
            state=JobApplicationWorkflow.STATE_PROCESSING,
            to_company=origin_company,
            eligibility_diagnosis=EligibilityDiagnosisMadeBySiaeFactory(),
        )
        eligibility_diagnosis_pk = job_application.eligibility_diagnosis.pk
        job_application.transfer_to(origin_user, target_company)
        job_application.refresh_from_db()

        assert job_application.to_company == target_company
        assert job_application.state == JobApplicationWorkflow.STATE_NEW
        assert job_application.eligibility_diagnosis is None
        assert not EligibilityDiagnosis.objects.filter(pk=eligibility_diagnosis_pk)

    def test_transfer_to_without_sender(self):
        origin_company = CompanyFactory(with_membership=True)
        target_company = CompanyFactory(with_membership=True)
        origin_user = origin_company.members.first()
        target_company.members.first()
        target_company.members.add(origin_user)

        job_application = JobApplicationFactory(
            to_company=origin_company,
            sent_by_authorized_prescriber_organisation=True,
            state=JobApplicationWorkflow.STATE_PROCESSING,
        )
        # Sender user account is deleted.
        job_application.sender = None
        job_application.save(update_fields=["sender"])

        job_application.transfer_to(origin_user, target_company)
        job_application.refresh_from_db()

        assert job_application.to_company == target_company
        assert job_application.state == JobApplicationWorkflow.STATE_NEW

    def test_model_fields(self):
        # Check new fields in model
        origin_company = CompanyFactory(with_membership=True)
        target_company = CompanyFactory(with_membership=True)

        origin_user = origin_company.members.first()
        target_user = target_company.members.first()
        target_company.members.add(origin_user)

        job_application = JobApplicationSentByCompanyFactory(
            state=JobApplicationWorkflow.STATE_PROCESSING,
            to_company=origin_company,
            eligibility_diagnosis=EligibilityDiagnosisMadeBySiaeFactory(),
            answer="Answer to job seeker",
            answer_to_prescriber="Answer to prescriber",
        )

        # Failing to transfer must not update new fields
        with pytest.raises(ValidationError):
            job_application.transfer_to(target_user, target_company)
        assert job_application.transferred_by is None
        assert job_application.transferred_from is None
        assert job_application.transferred_at is None

        with self.assertNumQueries(
            2  # Check user is in both origin and dest siae
            + 6  # Caused by `full_clean()` : `clean_fields()`
            + 3  # Integrity constraints check (full clean)
            + 1  # Update job application
            + 1  # Check if approvals are linked to diagnosis because of on_delete=set_null
            + 1  # Check if job applications are linked because of on_delete=set_null
            + 2  # Delete diagnosis and criteria made by the SIAE
            + 1  # Select user for email
        ):
            job_application.transfer_to(origin_user, target_company)

        job_application.refresh_from_db()

        assert job_application.transferred_by == origin_user
        assert job_application.transferred_from == origin_company
        assert timezone.localdate() == job_application.transferred_at.date()
        assert job_application.to_company == target_company
        assert job_application.state == JobApplicationWorkflow.STATE_NEW
        assert job_application.eligibility_diagnosis is None
        assert job_application.answer == ""
        assert job_application.answer_to_prescriber == ""

    def test_workflow_transitions(self):
        # `source` contains possible entry points of transition
        for from_state in JobApplicationWorkflow.transitions["transfer"].source:
            with self.subTest(from_state):
                job_application = JobApplicationSentByCompanyFactory(state=from_state)
                job_application.state = JobApplicationWorkflow.STATE_NEW
                job_application.save()  # Triggers transition check

    def test_transfer_must_notify_siae_and_job_seeker(self):
        # Send email notification of transfer to :
        # - origin SIAE
        # - job seeker
        # - Prescriber (if any linked eligibility diagnosis was not sent by a SIAE)
        origin_company = CompanyFactory(with_membership=True)
        target_company = CompanyFactory(with_membership=True)

        origin_user = origin_company.members.first()
        target_company.members.add(origin_user)

        job_application = JobApplicationSentByCompanyFactory(
            state=JobApplicationWorkflow.STATE_PROCESSING,
            to_company=origin_company,
            eligibility_diagnosis=EligibilityDiagnosisMadeBySiaeFactory(),
        )
        job_seeker = job_application.job_seeker

        job_application.transfer_to(origin_user, target_company)

        # Eligigibility diagnosis is done by SIAE : must not send an email
        assert len(mail.outbox) == 2

        assert len(mail.outbox[0].to) == 1
        assert origin_user.email in mail.outbox[0].to
        assert f"La candidature de {job_seeker.get_full_name()} a été transférée" == mail.outbox[0].subject
        assert "a transféré la candidature de :" in mail.outbox[0].body

        assert len(mail.outbox[1].to) == 1
        assert job_application.job_seeker.email in mail.outbox[1].to
        assert "Votre candidature a été transférée à une autre structure" in mail.outbox[1].subject
        assert "a transféré votre candidature à la structure" in mail.outbox[1].body

    def test_transfer_must_notify_prescriber(self):
        # Same test and conditions as above, but this time prescriber
        # at the origin of the eligibility disgnosis must be notified
        origin_company = CompanyFactory(with_membership=True)
        target_company = CompanyFactory(with_membership=True)

        origin_user = origin_company.members.first()
        target_company.members.add(origin_user)

        # Eligibility diagnosis was made by a prescriber
        job_application = JobApplicationSentByPrescriberFactory(
            state=JobApplicationWorkflow.STATE_PROCESSING,
            to_company=origin_company,
            eligibility_diagnosis=EligibilityDiagnosisFactory(),
        )
        job_seeker = job_application.job_seeker

        job_application.transfer_to(origin_user, target_company)

        assert len(mail.outbox) == 3

        # Other email content have been checked in previous test
        # Focusing on prescriber email content
        assert len(mail.outbox[2].to) == 1
        assert job_application.sender.email in mail.outbox[2].to
        assert f"La candidature de {job_seeker.get_full_name()} a été transférée" == mail.outbox[2].subject
        assert "a transféré la candidature de :" in mail.outbox[2].body

    def test_transfer_notifications_to_many_employers(self):
        # Same as test_transfer_must_notify_siae_and_job_seeker
        # but with to recipients for SIAE transfer notification
        origin_company = CompanyWith2MembershipsFactory()
        target_company = CompanyFactory(with_membership=True)

        origin_user_1 = origin_company.members.all()[0]
        origin_user_2 = origin_company.members.all()[1]
        target_company.members.add(origin_user_1)

        job_application = JobApplicationSentByCompanyFactory(
            state=JobApplicationWorkflow.STATE_PROCESSING,
            to_company=origin_company,
            eligibility_diagnosis=EligibilityDiagnosisMadeBySiaeFactory(),
        )
        job_seeker = job_application.job_seeker

        job_application.transfer_to(origin_user_1, target_company)

        # Only checking SIAE email
        assert len(mail.outbox) == 2
        assert len(mail.outbox[0].to) == 2
        assert origin_user_1.email in mail.outbox[0].to
        assert origin_user_2.email in mail.outbox[0].to
        assert f"La candidature de {job_seeker.get_full_name()} a été transférée" == mail.outbox[0].subject
        assert "a transféré la candidature de :" in mail.outbox[0].body
