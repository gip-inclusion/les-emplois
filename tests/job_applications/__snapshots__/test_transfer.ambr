# serializer version: 1
# name: test_model_fields
  dict({
    'num_queries': 14,
    'queries': list([
      dict({
        'origin': list([
          'Company.has_member[common_apps/organizations/models.py]',
          'JobApplication.can_be_transferred[job_applications/models.py]',
          'JobApplication.transfer[job_applications/models.py]',
        ]),
        'sql': '''
          SELECT %s AS "a"
          FROM "users_user"
          INNER JOIN "companies_companymembership" ON ("users_user"."id" = "companies_companymembership"."user_id")
          WHERE ("companies_companymembership"."id" IN
                   (SELECT U0."id"
                    FROM "companies_companymembership" U0
                    INNER JOIN "users_user" U2 ON (U0."user_id" = U2."id")
                    WHERE (U0."company_id" = %s
                           AND U2."is_active"
                           AND U0."is_active"))
                 AND "users_user"."id" = %s)
          LIMIT 1
        ''',
      }),
      dict({
        'origin': list([
          'Company.has_member[common_apps/organizations/models.py]',
          'JobApplication.can_be_transferred[job_applications/models.py]',
          'JobApplication.transfer[job_applications/models.py]',
        ]),
        'sql': '''
          SELECT %s AS "a"
          FROM "users_user"
          INNER JOIN "companies_companymembership" ON ("users_user"."id" = "companies_companymembership"."user_id")
          WHERE ("companies_companymembership"."id" IN
                   (SELECT U0."id"
                    FROM "companies_companymembership" U0
                    INNER JOIN "users_user" U2 ON (U0."user_id" = U2."id")
                    WHERE (U0."company_id" = %s
                           AND U2."is_active"
                           AND U0."is_active"))
                 AND "users_user"."id" = %s)
          LIMIT 1
        ''',
      }),
      dict({
        'origin': list([
          'JobApplication.save[<site-packages>/django/db/models/base.py]',
          'JobApplication.transfer[job_applications/models.py]',
        ]),
        'sql': '''
          UPDATE "job_applications_jobapplication"
          SET "eligibility_diagnosis_id" = NULL
          WHERE "job_applications_jobapplication"."id" = %s
        ''',
      }),
      dict({
        'origin': list([
          'JobApplication.transfer[job_applications/models.py]',
        ]),
        'sql': '''
          SELECT "job_applications_jobapplication"."id"
          FROM "job_applications_jobapplication"
          WHERE "job_applications_jobapplication"."eligibility_diagnosis_id" IN (%s)
          ORDER BY "job_applications_jobapplication"."created_at" DESC
        ''',
      }),
      dict({
        'origin': list([
          'JobApplication.transfer[job_applications/models.py]',
        ]),
        'sql': '''
          SELECT "approvals_approval"."id"
          FROM "approvals_approval"
          WHERE "approvals_approval"."eligibility_diagnosis_id" IN (%s)
          ORDER BY "approvals_approval"."created_at" DESC
        ''',
      }),
      dict({
        'origin': list([
          'JobApplication.transfer[job_applications/models.py]',
        ]),
        'sql': '''
          DELETE
          FROM "eligibility_selectedadministrativecriteria"
          WHERE "eligibility_selectedadministrativecriteria"."eligibility_diagnosis_id" IN (%s)
        ''',
      }),
      dict({
        'origin': list([
          'JobApplication.transfer[job_applications/models.py]',
        ]),
        'sql': '''
          DELETE
          FROM "eligibility_eligibilitydiagnosis"
          WHERE "eligibility_eligibilitydiagnosis"."id" IN (%s)
        ''',
      }),
      dict({
        'origin': list([
          'JobApplication.transfer[job_applications/models.py]',
        ]),
        'sql': '''
          SELECT "users_user"."id",
                 "users_user"."password",
                 "users_user"."last_login",
                 "users_user"."is_superuser",
                 "users_user"."username",
                 "users_user"."first_name",
                 "users_user"."last_name",
                 "users_user"."is_staff",
                 "users_user"."is_active",
                 "users_user"."date_joined",
                 "users_user"."address_line_1",
                 "users_user"."address_line_2",
                 "users_user"."post_code",
                 "users_user"."city",
                 "users_user"."department",
                 "users_user"."coords",
                 "users_user"."geocoding_score",
                 "users_user"."geocoding_updated_at",
                 "users_user"."ban_api_resolved_address",
                 "users_user"."insee_city_id",
                 "users_user"."title",
                 "users_user"."email",
                 "users_user"."phone",
                 "users_user"."kind",
                 "users_user"."identity_provider",
                 "users_user"."has_completed_welcoming_tour",
                 "users_user"."created_by_id",
                 "users_user"."external_data_source_history",
                 "users_user"."last_checked_at",
                 "users_user"."public_id",
                 "users_user"."address_filled_at",
                 "users_user"."first_login"
          FROM "users_user"
          INNER JOIN "companies_companymembership" ON ("users_user"."id" = "companies_companymembership"."user_id")
          WHERE "companies_companymembership"."id" IN
              (SELECT U0."id"
               FROM "companies_companymembership" U0
               INNER JOIN "users_user" U2 ON (U0."user_id" = U2."id")
               WHERE (U0."company_id" = %s
                      AND U2."is_active"
                      AND U0."is_active"))
          ORDER BY RANDOM() ASC
        ''',
      }),
      dict({
        'origin': list([
          'JobApplicationTransferredForEmployerNotification.should_send[communications/dispatch/base.py]',
          'JobApplicationTransferredForEmployerNotification.send[communications/dispatch/email.py]',
          'JobApplication.transfer[job_applications/models.py]',
        ]),
        'sql': '''
          SELECT %s AS "a"
          FROM "communications_notificationsettings"
          INNER JOIN "communications_disablednotification" ON ("communications_notificationsettings"."id" = "communications_disablednotification"."settings_id")
          INNER JOIN "communications_notificationrecord" ON ("communications_disablednotification"."notification_record_id" = "communications_notificationrecord"."id")
          WHERE ("communications_notificationsettings"."user_id" = %s
                 AND "communications_notificationsettings"."structure_pk" = %s
                 AND "communications_notificationsettings"."structure_type_id" = %s
                 AND "communications_notificationrecord"."notification_class" = %s)
          LIMIT 1
        ''',
      }),
      dict({
        'origin': list([
          'Email.save[<site-packages>/django/db/models/base.py]',
          'AsyncEmailBackend.send_messages[emails/tasks.py]',
          'JobApplicationTransferredForEmployerNotification.send[communications/dispatch/email.py]',
          'JobApplication.transfer[job_applications/models.py]',
        ]),
        'sql': '''
          INSERT INTO "emails_email" ("to",
                                      "cc",
                                      "bcc",
                                      "subject",
                                      "body_text",
                                      "from_email",
                                      "reply_to",
                                      "created_at",
                                      "esp_response")
          VALUES (%s::citext[], %s::citext[], %s::citext[], %s, %s, %s, %s::citext[], %s, %s) RETURNING "emails_email"."id"
        ''',
      }),
      dict({
        'origin': list([
          'JobApplicationTransferredForJobSeekerNotification.should_send[communications/dispatch/base.py]',
          'JobApplicationTransferredForJobSeekerNotification.send[communications/dispatch/email.py]',
          'JobApplication.transfer[job_applications/models.py]',
        ]),
        'sql': '''
          SELECT %s AS "a"
          FROM "communications_notificationsettings"
          INNER JOIN "communications_disablednotification" ON ("communications_notificationsettings"."id" = "communications_disablednotification"."settings_id")
          INNER JOIN "communications_notificationrecord" ON ("communications_disablednotification"."notification_record_id" = "communications_notificationrecord"."id")
          WHERE ("communications_notificationsettings"."user_id" = %s
                 AND "communications_notificationsettings"."structure_pk" IS NULL
                 AND "communications_notificationsettings"."structure_type_id" IS NULL
                 AND "communications_notificationrecord"."notification_class" = %s)
          LIMIT 1
        ''',
      }),
      dict({
        'origin': list([
          'Email.save[<site-packages>/django/db/models/base.py]',
          'AsyncEmailBackend.send_messages[emails/tasks.py]',
          'JobApplicationTransferredForJobSeekerNotification.send[communications/dispatch/email.py]',
          'JobApplication.transfer[job_applications/models.py]',
        ]),
        'sql': '''
          INSERT INTO "emails_email" ("to",
                                      "cc",
                                      "bcc",
                                      "subject",
                                      "body_text",
                                      "from_email",
                                      "reply_to",
                                      "created_at",
                                      "esp_response")
          VALUES (%s::citext[], %s::citext[], %s::citext[], %s, %s, %s, %s::citext[], %s, %s) RETURNING "emails_email"."id"
        ''',
      }),
      dict({
        'origin': list([
          'JobApplication.save[<site-packages>/django/db/models/base.py]',
        ]),
        'sql': '''
          UPDATE "job_applications_jobapplication"
          SET "job_seeker_id" = %s,
              "eligibility_diagnosis_id" = NULL,
              "geiq_eligibility_diagnosis_id" = NULL,
              "create_employee_record" = %s,
              "resume_link" = %s,
              "sender_id" = %s,
              "sender_kind" = %s,
              "sender_company_id" = %s,
              "sender_prescriber_organization_id" = NULL,
              "to_company_id" = %s,
              "state" = %s,
              "archived_at" = NULL,
              "archived_by_id" = NULL,
              "hired_job_id" = NULL,
              "message" = %s,
              "answer" = %s,
              "answer_to_prescriber" = %s,
              "refusal_reason" = %s,
              "refusal_reason_shared_with_job_seeker" = %s,
              "hiring_start_at" = %s,
              "hiring_end_at" = %s,
              "hiring_without_approval" = %s,
              "origin" = %s,
              "approval_id" = NULL,
              "approval_delivery_mode" = %s,
              "approval_number_sent_by_email" = %s,
              "approval_number_sent_at" = NULL,
              "approval_manually_delivered_by_id" = NULL,
              "approval_manually_refused_by_id" = NULL,
              "approval_manually_refused_at" = NULL,
              "transferred_at" = %s,
              "transferred_by_id" = %s,
              "transferred_from_id" = %s,
              "created_at" = %s,
              "updated_at" = %s,
              "processed_at" = NULL,
              "prehiring_guidance_days" = NULL,
              "contract_type" = %s,
              "nb_hours_per_week" = NULL,
              "contract_type_details" = %s,
              "qualification_type" = %s,
              "qualification_level" = %s,
              "planned_training_hours" = NULL,
              "inverted_vae_contract" = NULL,
              "diagoriente_invite_sent_at" = NULL
          WHERE "job_applications_jobapplication"."id" = %s
        ''',
      }),
      dict({
        'origin': list([
          'JobApplicationTransitionLog.save[<site-packages>/django/db/models/base.py]',
        ]),
        'sql': '''
          INSERT INTO "job_applications_jobapplicationtransitionlog" ("transition",
                                                                      "from_state",
                                                                      "to_state",
                                                                      "timestamp",
                                                                      "job_application_id",
                                                                      "user_id",
                                                                      "target_company_id")
          VALUES (%s, %s, %s, %s, %s, %s, %s) RETURNING "job_applications_jobapplicationtransitionlog"."id"
        ''',
      }),
    ]),
  })
# ---
